include/swift/Syntax/SyntaxData.h

```diff
@@ -76,19 +76,29 @@ class SyntaxData {
   const AbsoluteRawSyntax AbsoluteRaw;
   RC<RefCountedBox<SyntaxData>> Parent;
 
+  /// If this node is the root of a Syntax tree (i.e. \c Parent is \c nullptr ),
+  /// the arena in which this node's \c RawSyntax node has been allocated.
+  /// This keeps this \c RawSyntax nodes referenced by this tree alive.
+  RC<SyntaxArena> Arena;
+
+  /// Create a intermediate node with a parent.
   SyntaxData(AbsoluteRawSyntax AbsoluteRaw,
              const RC<RefCountedBox<SyntaxData>> &Parent)
-      : AbsoluteRaw(AbsoluteRaw), Parent(Parent) {}
+      : AbsoluteRaw(AbsoluteRaw), Parent(Parent), Arena(nullptr) {}
+
+  /// Create a new root node
+  SyntaxData(AbsoluteRawSyntax AbsoluteRaw, const RC<SyntaxArena> &Arena)
+      : AbsoluteRaw(AbsoluteRaw), Parent(nullptr), Arena(Arena) {}
```

Comment:
I think 
```cpp
SyntaxData(AbsoluteRawSyntax AbsoluteRaw)
  : AbsoluteRaw(AbsoluteRaw), Arena(AbsoluteRaw.getRaw()->getArena()) {}
```
is safer *if* there's no need to refer a different `Arena`.

Reply:
You are absolutely right!